### PR TITLE
Migrated Header to Headers

### PR DESCRIPTION
**Pull request by oldergod**
`Header` only lives inside `Hpack` now (because of the dynamic table thing)
Everything else is `Headers`

- [ ] Where to move the `Headers` listener? Inside Http2Stream? As itself within the `internal.okhttp2` package?
- [ ] Are we cool keeping `Header` for the dynamic table inside `Hpack`? Do we want to refactor that too?

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/Headers.java [hunk 1]
```diff
@@ -48,6 +49,15 @@
...
  */
 public final class Headers {
+  // TODO(oldergod) make them String ?
+  // Special header names defined in HTTP/2 spec.
+  public static final ByteString PSEUDO_PREFIX = ByteString.encodeUtf8(":");
+  public static final ByteString RESPONSE_STATUS = ByteString.encodeUtf8(":status");
+  public static final ByteString TARGET_METHOD = ByteString.encodeUtf8(":method");
+  public static final ByteString TARGET_PATH = ByteString.encodeUtf8(":path");
+  public static final ByteString TARGET_SCHEME = ByteString.encodeUtf8(":scheme");
+  public static final ByteString TARGET_AUTHORITY = ByteString.encodeUtf8(":authority");
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Headers.java]: Don't think these make sense here. This API works on strings, not byte strings. Plus the values of these are never exposed publicly. What about on the `Http2` class?

File: okhttp/src/main/java/okhttp3/Headers.java [hunk 2]
```diff
@@ -275,6 +285,21 @@ static void checkValue(String value, String name) {
     }
   }
 
+  public interface Listener {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Headers.java]: Doesn't seem like it should public API

File: okhttp/src/main/java/okhttp3/Headers.java [hunk 3]
```diff
@@ -275,6 +285,21 @@ static void checkValue(String value, String name) {
...
+    void onHeaders(List<Headers> headers);
+  }
+
+  //// TODO(oldergod) maybe delete me? maybe not
+  public static Headers flattenList(List<Headers> headersBlocks) {
+    Builder builder = new Builder();
+    for (Headers headers : headersBlocks) {
+      if (headers == null) throw new IllegalStateException("WTF");
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Headers.java]: Should be NPE and have a meaningful message

File: okhttp/src/main/java/okhttp3/Headers.java (hunk 2 above)
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/Headers.java]: It's used inside `Http2Stream` to forward received headers blocks.

File: okhttp/src/main/java/okhttp3/Headers.java (hunk 1 above)
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/Headers.java]: Looking at this now. There is some places there are `String` comparison, sometimes `ByteString` comparison; trying to find the right line to draw.

File: okhttp/src/main/java/okhttp3/Headers.java (hunk 2 above)
Comment (JakeWharton) [reply okhttp/src/main/java/okhttp3/Headers.java]: That's an internal class. This callback is unused by public API.

File: okhttp/src/main/java/okhttp3/Headers.java (hunk 1 above)
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/Headers.java]: Gave it a shot under Http2; still had to be public to be used inside `MockWebServer`.

File: okhttp/src/main/java/okhttp3/Headers.java (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Headers.java]: We're trying to figure out where to put it for duplex calls, when users will be able to subscribe to headers updates.

File: okhttp/src/main/java/okhttp3/internal/http2/Hpack.java [hunk 4]
```diff
@@ -32,7 +31,7 @@
 /**
  * Read and write HPACK v10.
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http2/Hpack.java]: I assume the v10 was a draft

```suggestion
 * Read and write HPACK.
```

**Pull request closed without merging**